components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -2550,9 +2553,8 @@ subroutine cldliq_immersion_freezing(t,lamc,mu_c,cdist1,qc_incld,    &
 
    if (qc_incld.ge.qsmall .and. t.le.rainfrze) then
       ! for future: calculate gamma(mu_c+4) in one place since its used multiple times  !AaronDonahue, TODO
-      dum   = (1._rtype/lamc)**3
-      Q_nuc = cons6*cdist1*gamma(7._rtype+mu_c)*exp(aimm*(zerodegc-t))*dum**2
-      N_nuc = cons5*cdist1*gamma(mu_c+4._rtype)*exp(aimm*(zerodegc-t))*dum
+      Q_nuc = cons6*cdist1*bfb_gamma(7._rtype+mu_c)*bfb_exp(aimm*(zerodegc-t))*bfb_pow(lamc, -6.0)
```

Comment:
We don't want to use pow for integer exponentiation. Not sure if this is done elsewhere, but in any case, it shouldn't be. I'm pretty sure `x**3` in F90 is equivalent to `x*x*x`, so C++ should use a `cube` function here that does that. Then separately squaring it, again with integer 2, ie `x*x`, makes sense. So I think the original formulation is fine. If we want to be clear about saved work, we could save `bfb_exp(aimm*(zerodegc-t))` to a temp and reuse in each formula. The compiler may or may not do that already.

Reply:
Yeah, I was wondering about integer exponentiation. I'll check to see whether I've done it in other conversions.

I'll reinstate the temporary, and I could introduce another to hold the argument to bfb_exp.

Reply:
Instead of a `cxx_cube` function, though, wouldn't it be nice to have a `cxx_ipow` or `cxx_powi` function that takes an integer argument for the exponent and uses the corresponding overloaded C++ `pow`? That seems a bit more flexible.

Reply:
Sure, however you want to design it. `pow<3>(x)` in C++ would be great.

Reply:
I had added an ipow_scalar last year but I removed it a few months ago and I can't remember why. I think it was causing BFB problems between CPU and GPU. see commit c16a69fd15295d8dfa5ececaa7656922f2fabae1

Reply:
In hommexx I also had to impl an int pow function for the bfb pow impl.

[See here](https://github.com/E3SM-Project/E3SM/blob/5a6c92b8f09a38dde98688a0f17adcbcc2bc51b1/components/homme/src/share/cxx/utilities/BfbUtils.hpp#L64)

The funciton is not templated, cause my exponent was a runtime var, but we could write both a template and runtime version.

Reply:
Since you are bridging to ipow directly from micro_p3.F90, I think it should work.

Reply:
The way I have it seems to work on the Fortran end for now. I think we can probably optimize it on the C++ end as needed. Is this okay for the time being?

Reply:
@jgfouca is it possible the unit test in the commit you pointed to failed only when the exponent was <0? Here we're focused on >0 exponents. For <0, perhaps CPU and GPU would have agreed if division were done at the end rather than the start.

By the way, rather than use one ipow function with a loop, I recommend a templated one with a default impl of a loop but specializations for the common cases, such as 3.

Reply:
@ambrad , possible. it's been a long time.

Reply:
I vaguely remember, the only way to ensure BFB integer pow was to unroll the pow into simple multiplies. For example:

```
x**2
```

becomes
```
x*x
```

If the exponent is a variable, it's possible things will work better since there won't be as much compiler trickery going on. If the compiler sees a literal exponent (eg 2 or 3) it can get tempted (on either fortran or C++ side) to do subtle optimizations.

Reply:
I'm fine with whatever is done. But just tbc, we're interested in BFB only in the case of no optimizations, our debug builds. So I think the point stands that F90 uses unrolling in the case of integer exponents, perhaps up to some magnitude, and in C++ we can write our own. In a debug build, the compiler can't do anything to change these.

Reply:
>  no optimizations, our debug builds

I agree. I was surprised to see that my compiler was doing the optimizations I described, even in debug.

Reply:
At least on some platforms, I think this issue may have been on intel.

Reply:
It seems like we could avoid bit-for-bit issues by adding `cxx_square` and `cxx_cube` functions for x**2 and x**3 cases, exposed here as `bfb_square` and `bfb_cube`. I'm willing to do that if no one objects.